Cli

EllyBax opened the pull request:
this pull request intoduces a fleshed out cli part of the evkit project with al the main commands working and in simple need of refactoring

EllyBax committed:
Web init

EllyBax committed:
feat(web): added shadcn components to get started

EllyBax committed:
feat(utils): initial encryption setup

EllyBax committed:
fix(db): all actions now correctly use node

FEAT: Added encryption and decryption algorithms

EllyBax committed:
feat(web): AUTH added clerk and convex providers to get started with auth

EllyBax committed:
feat(web) : SIGNIN component added

EllyBax committed:
Build packages before consuming in apps
Fixed tsconfig and moved files into src/ for consistecncy

EllyBax committed:
fix (code review): applied refactors and fixes to ensure security and perfomance

EllyBax committed:
feat(encrytion): drafted version 1 working encryption classes

EllyBax committed:
feat(encryption): V1 complete

EllyBax committed:
fix(build): fixed the build script for the db to ensure it's built for the web and cli to consume

EllyBax committed:
feat(db): Added alias and postbuild script to mitigate errors

TODO: the fix isn't working for web still though

EllyBax committed:
Initialized CLI

EllyBax committed:
feat(db): BREAKING CHANGE always pass the salts and hashes, take that responsibility away from the db

EllyBax committed:
fear(web): CLI authentication flow on the web side completed

EllyBax committed:
feat(cli): first command done 90%

EllyBax committed:
feat(cli): added gitignore and db test

EllyBax committed:
feat(auth): added auth commands to the cli

NOT FULLY TESTED YET, due to db erroring

EllyBax committed:
feat(auth): added whoami command and modified logout to revoke db session too

EllyBax committed:
feat(INITCMD): scaffolded the initialization command

EllyBax committed:
refactor(init): refactored the init with helpers and added audit for blame

EllyBax committed:
feat(web): added convex to the web

EllyBax committed:
fixed web db imports

EllyBax committed:
Fixed cli db access

EllyBax committed:
feat(push): added addVars method to dbAPI and implemented the push command

EllyBax committed:
fix: fixed fetching variables by branch and deletedat

EllyBax committed:
in-fix: why isn't decryptworking??

EllyBax committed:
in-fix: reverted to optional branch

EllyBax committed:
feat: modified db snapshots

EllyBax committed:
feat(db): rethought project snapshots

EllyBax committed:
feat(hash-vars): added hash for var comparison

EllyBax committed:
Feat(hash): fixed hashing with web crypto

EllyBax committed:
Fix(path): fixed gitignore file declaration

EllyBax committed:
feat:(sync) initial implementation of sync

EllyBax committed:
feat:(env) ensured to always consolidate vars to .env.local

EllyBax committed:
feat(tests): added simple unit tests for cli and db

EllyBax committed:
feat(get): implemented the get command

EllyBax committed:
feat: Implement set and delete commands and fix decryption bug

This commit introduces several new features and a bug fix.

Features:
- `set` command: A new `set` command has been added to the CLI to allow creating and updating single environment variables.
- `delete` command: A new `delete` command has been added to the CLI to allow deleting single environment variables.
- The `set` and `delete` commands are also implemented on the database level with new mutations in Convex.

Bug Fixes:
- A bug in the `decryptVariable` function that caused a crash when trying to decrypt an empty string has been fixed. This occurred when a variable was "deleted" by setting its value to an empty string.
- The `sync` commands error handling has been improved.
- The `pull` command now correctly updates the local environment file hash after pulling changes.

Other Changes:
- The CLI now displays a welcome message on startup.
- The `dotenv` configuration has been updated to explicitly use `.env.local`.
- The database schema for variables has been updated to track which user last updated a variable.

EllyBax committed:
feat: Implemented unlink command

this commit renames the commads/ dir to commands/ and add a new command unlink to delete a project and it's variables

the init command also now confirms with the user prior to consolidating the envs to .env.local

jrdev-org committed:
refactor: fetching the cli version from the constants and displaying it in the initial command

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,797 @@
+  return linkedProject;
+}
+
+async function getProjects(dir: string) {
+  const dirContents = await fs.stat(PROJECTS_DIR).catch(() => null);
+  if (!dirContents || !dirContents.isDirectory()) return false;
+
+  const projects = await fs.readdir(PROJECTS_DIR);
+  if (!projects.length) return false;
+  return projects.filter((p) =>
+    p.includes(dir)
+  ) as `projectName-projectStage`[];
+}
```
This is great, thanks!

jrdev-org committed:
Update apps/cli/src/commands/actions.ts

Review project name matching logic.

Line 127-129 filters projects using .includes(dir), which will match partial directory names. For example, if dir is "app", it will match both "app-dev" and "myapp-dev"

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: Critical bug: Early return prevents token storage.

If the device is already registered (line 74 returns "found"), the function returns early on line 75, skipping token storage and session initialization. This leaves the user in a broken state where authentication appears to succeed but no token is stored.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: Stop auth server on successful authentication.

The auth server is stopped in the error path but not on the success path. This leaves the server running and the port bound after successful authentication, which could cause issues if the user tries to authenticate again quickly or runs multiple instances.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on apps/cli/src/commands/init.ts:
```diff
@@ -0,0 +1,345 @@
+  }
+}
+
+export async function getEnvFileHash(filePath: string): Promise<string> {
+  const envVars = await loadEnvFile(filePath);
+  if (Object.keys(envVars).length === 0) {
+    return "";
+  }
+  const canonical = Object.entries(envVars)
+    .sort(([keyA], [keyB]) => keyA.localeCompare(keyB))
+    .map(([key, value]) => `${key}=${value}`)
+    .join("\n");
+  const hashBuffer = await crypto.subtle.digest(
+    "SHA-256",
+    new TextEncoder().encode(canonical)
+  );
+  const hashArray = Array.from(new Uint8Array(hashBuffer));
+  const hashHex = hashArray
+    .map((b) => b.toString(16).padStart(2, "0"))
+    .join("");
+  return hashHex;
+}
```
Fixed

jrdev-org committed:
fix: cross-platform path handling.

Using .split("/").pop() to extract the project name will fail on Windows, which uses backslashes. Use path.basename() for cross-platform compatibility

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on packages/db/src/encryption.ts:
```diff
@@ -0,0 +1,162 @@
+    return { userId: team.ownerId, team };
+  }
+
+  async createSalt(): Promise<string> {
+    // TODO: add a limit on the number of salts per team
+    const { userId, team } = await this.getUserAndTeam();
+    const existing = await convex.query(api.salts.get, { teamId: team._id });
+    if (existing.length > 0) {
+      return existing[0].salt;
+    }
+    const salt = VariableEncryption.generateSalt();
+    await convex.mutation(api.salts.create, {
+      teamId: team._id,
+      salt,
+      callerId: userId,
+    });
+    return salt;
+  }
```
fixed: will be available on next commit

jrdev-org committed:
fix: Non-null assertion on CONVEX_URL will throw at runtime if missing.

The non-null assertion operator (!) will cause the ConvexHttpClient constructor to receive undefined if CONVEX_URL is missing, leading to a runtime error. Validate the environment variable presence and provide a clear error message.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on packages/db/tsconfig.json:
```diff
@@ -1,9 +1,16 @@
   "extends": "../../tsconfig.base.json",
-  "include": ["**/*.ts"],
-  "exclude": ["dist"],
+  "include": ["src/**/*.ts", "convex/**/*.ts", "convex/**/*.js"],
+  "exclude": ["node_modules"],
```
fix applied locally, change will be available in next push

jrdev-org committed:
Fixed the import path typo.

The import path contains a typo: "commads" should be "commands". This will cause a module resolution error at runtime.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: Potential race condition in test.

The test creates an HTTP server but doesn't wait for it to be fully listening before calling pickAvailablePort(). This could lead to flaky test behavior where the port appears available when it should be occupied.

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: readLinkedProject has the same partial matching issue.

Lines 134-135 use .includes() twice for matching, which will have false positives similar to getProjects

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: serialization of env variables before writing to the env file

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org committed:
fix: sanitize project names to avoid errors if inputs have special chars

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on apps/cli/src/commands/init.ts:
```diff
@@ -0,0 +1,345 @@
+  });
+}
+
+export async function updateProjectsDir(project: Project, hash: string) {
+  const filePath = path.join(PROJECTS_DIR, `${project.name}-${project.stage}`);
+
+  const enriched: LinkedProject = {
+    ...project,
+    linkedAt: Date.now(),
+    hash: hash,
+  };
+
+  await fs.writeFile(filePath, JSON.stringify(enriched, null, 2), {
+    encoding: "utf-8",
+    mode: 0o600, // secure owner-only read/write
+  });
+}
```
applied, this will be available pending the pull and push of the local code

jrdev-org commented on packages/db/convex/users.ts:
```diff
@@ -22,35 +27,37 @@ export const create = mutation({
       updatedAt: Date.now(),
     });
 
+    // 4) create team salt
+    await ctx.db.insert("salts", {
+      teamId: newTeamId,
+      salt: args.salt,
+    });
```
fixed, the salt is created strong via a helper function in ./helpers.js

jrdev-org commented on packages/db/src/index.ts:
```diff
@@ -0,0 +1,315 @@
+import { type Id } from "../convex/_generated/dataModel.js";
+import { ConvexHttpClient } from "convex/browser";
+import dotenv from "dotenv";
+dotenv.config({
+  path: process.cwd() + "/.env.local",
+});
```
fixed locally

jrdev-org commented on packages/utils/src/index.ts:
```diff
@@ -0,0 +1,139 @@
+  /**
+   * Encrypt a message using the user's salt
+   */
+  static encryptMessage(message: string, userSalt: string): string {
+    try {
+      const key = this.deriveKey(userSalt);
+      const iv = crypto.randomBytes(this.IV_LENGTH);
+      const cipher = crypto.createCipheriv(this.ALGORITHM, key, iv);
+
+      let encrypted = cipher.update(message, "utf8", "hex");
+      encrypted += cipher.final("hex");
+
+      const authTag = cipher.getAuthTag();
+
+      // Combine IV + authTag + encrypted data
+      const combined =
+        iv.toString("hex") + ":" + authTag.toString("hex") + ":" + encrypted;
+      return combined;
+    } catch (error) {
+      throw new Error(
+        `Encryption failed: ${error instanceof Error ? error.message : String(error)}`
+      );
+    }
+  }
```
fix applied

jrdev-org committed:
fix: replaced duplicate code with a call to getLinkedProject

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jrdev-org commented on apps/cli/src/commands/init.ts:
```diff
@@ -0,0 +1,345 @@
+  });
+  if (!confirmLink) return log.info("Aborting...");
+
+  const projectSpinner = log.task("Linking project...").start();
+  const variables = await safeCall(() =>
+    dbApi.projects.getVars(projectToLink._id as Id<"projects">, hash)
+  )();
+  if ("error" in variables) throw new Error(variables.error);
+  if (!variables.changed) {
+    throw new Error("No changes in environment file. Nothing to link.");
+  }
```
fixed, locally

jrdev-org commented on packages/utils/src/index.ts:
```diff
@@ -0,0 +1,139 @@
+  /**
+   * Decrypt a message using the user's salt
+   */
+  static decryptMessage(encryptedData: string, userSalt: string): string {
+    try {
+      const parts = encryptedData.split(":");
+      if (parts.length !== 3) {
+        throw new Error("Invalid encrypted data format");
+      }
+
+      const iv = Buffer.from(parts[0], "hex");
+      const authTag = Buffer.from(parts[1], "hex");
+      const encrypted = parts[2];
+
+      const key = this.deriveKey(userSalt);
+      const decipher = crypto.createDecipheriv(this.ALGORITHM, key, iv);
+      decipher.setAuthTag(authTag);
+
+      let decrypted = decipher.update(encrypted, "hex", "utf8");
+      decrypted += decipher.final("utf8");
+
+      return decrypted;
+    } catch (error) {
+      throw new Error(
+        `Decryption failed: ${error instanceof Error ? error.message : String(error)}`
+      );
+    }
+  }
```
fixed

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,797 @@
+    log.debug(
+      `Variable ${chalk.bold(pulled)} found on the server with value ${chalk.bold(variables[pulled])}`
+    );
+    // log.debug(
+    //   `Variable ${chalk.bold(key)} not found locally. Run ${chalk.bold("envkit pull")} to pull from the cloud.`
+    // );
```
okay

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,797 @@
+        },
+      });
+
+      const hash = await getEnvFileHash(envFile);
+      await updateProjectsDir(res, hash);
```
that's fine,I'll rename res to updated project

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,797 @@
+    await runDelete(key, options.allowOverride, stage);
+  });
+
+export const deleteAllCmd = new Command("delete-all")
+  .description("Delete all variables")
+  .action(async () => {
+    log.debug(
+      `Unlink your project by running ${chalk.bold("envkit unlink")} to delete all variables`
+    );
+    process.exit(0);
+  });
```
thanks, let me make it an alias

jrdev-org commented on apps/cli/src/commands/projects.ts:
```diff
@@ -0,0 +1,79 @@
+    const deleteSpinner = log.task("Deleting project from database...").start();
+    const res = await safeCall(() =>
+      dbApi.projects.remove(
+        token.userId as unknown as Id<"users">,
```
fixed

jrdev-org commented on apps/cli/src/commands/projects.ts:
```diff
@@ -0,0 +1,79 @@
+    await fs.unlink(projectFilePath);
+
+    log.success(`Successfully unlinked project "${linkedProject.name}".`);
+  } else {
+    log.warn(
+      `To delete the project later, please do it from the web dashboard. Or re-run this command with the --force flag.`
+    );
+    process.exit(0);
+  }
```
fixed

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,764 @@
+  if ("error" in dbVars) throw new Error(dbVars.error);
+  if (!dbVars.changed) {
+    pushSpinner.succeed("Already up to date");
+    process.exit(0);
+  }
+
+  const encryptedVariables: { name: string; value: string }[] = [];
+  const teamService = new TeamService(
+    linkedProject.teamId,
+    token.userId as unknown as Id<"users">
+  );
+
+  await Promise.all(
+    Object.entries(variables).map(async ([k, v]) => {
+      const encrypted = await teamService.encryptVariable(v);
+      encryptedVariables.push({ name: k, value: encrypted });
+    })
+  );
+
+  const res = await safeCall(
+    async () =>
+      await dbApi.projects.addVars(
+        linkedProject._id,
+        token.userId as unknown as Id<"users">,
+        encryptedVariables
+      )
+  )();
+  if (!res || "error" in res) {
+    pushSpinner.fail("Error pushing variables!");
+    throw new Error(res?.error || "Unknown error");
+  }
+
+  const newHash = await getEnvFileHash(envFile);
+  await updateProjectsDir(res.updatedProject, newHash);
+
+  pushSpinner.succeed(
+    `Variables pushed successfully! ${res.additions.length} new, ${res.removals.length} removed, ${res.conflicts.length} modified.`
+  );
+  return;
+}
```
fixed

jrdev-org replied:
though I don't quite see the error

jrdev-org commented on apps/cli/src/commands/actions.ts:
```diff
@@ -0,0 +1,764 @@
+    ) {
+      action = "pull";
+    } else if (
+      localHash !== linkedProject.hash &&
+      serverHash !== linkedProject.hash
+    ) {
+      log.warn("Both local and cloud have diverged!");
+      const conflictResolution = await select({
+        message: "Resolve conflict",
+        choices: [
+          { name: "Pull (overwrite local)", value: "pull" },
+          { name: "Push (overwrite cloud)", value: "push" },
+          { name: "Abort", value: "abort" },
+        ],
+      });
+      if (conflictResolution === "abort") {
+        log.info("Sync aborted.");
+        process.exit(0);
+      }
+      action = conflictResolution as "pull" | "push";
+    } else {
+      log.success("Already in sync.");
+      process.exit(0);
+    }
+
+    const confirmAction = await confirm({
+      message: `Do you want to ${action} changes?`,
+      default: true,
+    });
+    if (!confirmAction) {
+      log.info("Sync aborted.");
+      process.exit(0);
+    }
+
+    if (action === "pull") await runPull(stage);
+    if (action === "push") await runPush(stage);
+
+    log.success("Sync complete!");
+    process.exit(0);
+  });
```
fixed too

jrdev-org commented on packages/db/src/index.ts:
```diff
@@ -0,0 +1,318 @@
+      });
+    },
+  },
+  variables: {
+    create: async (
+      projectId: Id<"projects">,
+      name: string,
+      encrypted: string,
+      branch?: string
+    ) => {
+      return await convex.mutation(api.variables.create, {
+        projectId,
+        name,
+        encryptedValue: encrypted,
+        branch,
+      });
+    },
+    update: async (
+      projectId: Id<"projects">,
+      name: string,
+      value: string,
+      branch?: string
+    ) => {
+      return await convex.mutation(api.variables.update, {
+        projectId,
+        name,
+        value,
+        branch,
+      });
+    },
```
fixed

jrdev-org commented:
Thanks to coderabbit, for the review all fixes applied now

jrdev-org commented:
continue with the good work Sir!

jrdev-org reviewed (commented):
Looks good! let's go on
